File: Data Structures & Algorithms/valid-binary-search-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        def dfs(node, interval):
            if node is None:
                return True
            if node and (interval[0] >= node.val or interval[1] <= node.val):
                return False
            checkLeft = dfs(node.left, [interval[0], node.val])
            checkRight = dfs(node.right, [node.val, interval[1]])
            return checkLeft and checkRight

        return dfs(root, [float("-inf"), float("inf")])

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-3.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
int maxPathSum(TreeNode *root)
    {
        int maxFound = INT_MIN;
        dfs(root, maxFound);
        return maxFound;
    }

    int dfs(TreeNode *root, int &maxFound)
    {
        if (!root)
        {
            return 0;
        }

        int left = dfs(root->left, maxFound);
        int right = dfs(root->right, maxFound);

        int maxChild = max(left, max(right, 0));
        int nPlusChild = root->val + maxChild;
        int split = left + right + root->val;
        maxFound = max({maxFound, split, nPlusChild, root->val});
        return nPlusChild;
    }
};
